File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/utils/GroundContactUtils.java
package com.github.aadvorak.artilleryonline.battle.utils;

import com.github.aadvorak.artilleryonline.battle.common.Constants;
import com.github.aadvorak.artilleryonline.battle.common.Contact;
import com.github.aadvorak.artilleryonline.battle.common.Position;
import com.github.aadvorak.artilleryonline.battle.common.lines.Circle;
import com.github.aadvorak.artilleryonline.battle.common.lines.HalfCircle;
import com.github.aadvorak.artilleryonline.battle.common.lines.Segment;
import com.github.aadvorak.artilleryonline.battle.common.lines.Trapeze;
import com.github.aadvorak.artilleryonline.battle.model.RoomModel;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class GroundContactUtils {

    public static Set<Contact> getGroundContacts(HalfCircle halfCircle, RoomModel roomModel, boolean withMaxDepth) {
        var groundIndexes = BattleUtils.getGroundIndexesBetween(halfCircle.center().getX() - halfCircle.radius(),
                halfCircle.center().getX() + halfCircle.radius(), roomModel);
        var contacts = new HashSet<Contact>();
        if (groundIndexes.isEmpty()) {
            return contacts;
        }
        var bottom = halfCircle.chord();
        var i = 0;
        var halfCircleNormal = halfCircle.center().vectorTo(halfCircle.center()
                .shifted(1.0, halfCircle.angle() + Math.PI / 2));
        var maxDepth = withMaxDepth ? roomModel.getSpecs().getGroundMaxDepth() : 0.0;
        while (i < groundIndexes.size()) {
            var position = BattleUtils.getGroundPosition(groundIndexes.get(i), roomModel);

            Contact bottomContact = getGroundContact(bottom, position, maxDepth, "HalfCircle bottom with ground");

            Contact topContact = null;
            var distance = halfCircle.center().distanceTo(position);
            var radiusVector = halfCircle.center().vectorTo(position);
            if (distance < halfCircle.radius() && halfCircleNormal.dotProduct(radiusVector) > 0) {
                var depth = halfCircle.radius() - distance - maxDepth;
                var contactPosition = halfCircle.center().shifted(halfCircle.radius(),
                        halfCircle.center().angleTo(position));
                topContact = Contact.withUncheckedDepthOf(
                        depth,
                        getGroundAngle(position, groundIndexes.get(i), roomModel),
                        contactPosition,
                        "HalfCircle top with ground"
                );
            }

            Contact resultContact = null;
            if (topContact != null && bottomContact != null) {
                resultContact = topContact.depth() < bottomContact.depth() ? topContact : bottomContact;
            } else if (topContact != null) {
                resultContact = topContact;
            }  else if (bottomContact != null) {
                resultContact = bottomContact;
            }

            if (resultContact != null && resultContact.depth() >= Constants.INTERPENETRATION_THRESHOLD) {
                contacts.add(resultContact);
            }

            i++;
        }
        return contacts;
    }

    public static Set<Contact> getGroundContacts(Trapeze trapeze, RoomModel roomModel, boolean withMaxDepth) {
        var maxDistance = trapeze.maxDistanceFromCenter();
        var groundIndexes = BattleUtils.getGroundIndexesBetween(trapeze.position().getX() - maxDistance,
                trapeze.position().getX() + maxDistance, roomModel);
        var contacts = new HashSet<Contact>();
        if (groundIndexes.isEmpty()) {
            return contacts;
        }
        var i = 0;
        var maxDepth = withMaxDepth ? roomModel.getSpecs().getGroundMaxDepth() : 0.0;
        var bottom = trapeze.bottom();
        var top = trapeze.top();
        var right = trapeze.right();
        var left = trapeze.left();
        while (i < groundIndexes.size()) {
            var position = BattleUtils.getGroundPosition(groundIndexes.get(i), roomModel);

            var bottomContact = getGroundContact(bottom, position, maxDepth, "Trapeze bottom with ground");
            var topContact = getGroundContact(top, position, maxDepth, "Trapeze top with ground");
            var rightContact = getGroundContact(right, position, maxDepth,  "Trapeze right with ground");
            var leftContact = getGroundContact(left, position, maxDepth, "Trapeze left with ground");

            Contact resultContact = null;
            if (topContact != null) {
                resultContact = topContact;
            }
            if (bottomContact != null && (resultContact == null || bottomContact.depth() < resultContact.depth())) {
                resultContact = bottomContact;
            }
            if (rightContact != null && (resultContact == null || rightContact.depth() < resultContact.depth())) {
                resultContact = rightContact;
            }
            if (leftContact != null && (resultContact == null || leftContact.depth() < resultContact.depth())) {
                resultContact = leftContact;
            }

            if (resultContact != null && resultContact.depth() >= Constants.INTERPENETRATION_THRESHOLD) {
                contacts.add(resultContact);
            }
            i++;
        }
        return contacts;
    }

    public static Contact getGroundContact(Circle circle, RoomModel roomModel, boolean withMaxDepth) {
        var groundIndexes = BattleUtils.getGroundIndexesBetween(circle.center().getX() - circle.radius(),
                circle.center().getX() + circle.radius(), roomModel);
        if (groundIndexes.isEmpty()) {
            return null;
        }
        Position nearestPosition = null;
        Double minimalDistance = null;
        Integer index = null;
        var i = 0;
        while (i < groundIndexes.size()) {
            var position = BattleUtils.getGroundPosition(groundIndexes.get(i), roomModel);
            var distance = position.distanceTo(circle.center());
            if (distance < circle.radius()) {
                if (minimalDistance == null || distance < minimalDistance) {
                    nearestPosition = position;
                    minimalDistance = distance;
                    index = groundIndexes.get(i);
                }
            }
            i++;
        }
        if (nearestPosition != null) {
            var depth = getGroundDepth(circle, nearestPosition.getY(), minimalDistance);
            if (withMaxDepth) depth -= roomModel.getSpecs().getGroundMaxDepth();
            var contactPosition = circle.center().shifted(circle.radius(),
                    circle.center().angleTo(nearestPosition));
            return Contact.of(
                    depth,
                    getGroundAngle(nearestPosition, index, roomModel),
                    contactPosition
            );
        }
        nearestPosition = BattleUtils.getNearestGroundPosition(circle.center().getX(), roomModel);
        var depth = getGroundDepth(circle, nearestPosition.getY());
        if (withMaxDepth) depth -= roomModel.getSpecs().getGroundMaxDepth();
        return Contact.of(
                depth,
                0.0,
                circle.center().shifted(circle.radius(), -Math.PI / 2)
        );
    }

    private static Contact getGroundContact(Segment segment, Position groundPosition, double maxDepth, String description) {
        var segmentPoint = segment.findPointWithX(groundPosition.getX());
        if (segmentPoint != null && segmentPoint.getY() < groundPosition.getY()) {
            var projection = GeometryUtils.getPointToSegmentProjection(groundPosition, segment);
            if (projection != null) {
                var depth = groundPosition.distanceTo(projection) - maxDepth;
                var normal = groundPosition.vectorTo(projection).normalized();
                return Contact.withUncheckedDepthOf(depth, normal, projection, description);
            }
        }
        return null;
    }

    private static double getGroundDepth(Circle circle, double y, double distance) {
        if (y <= circle.center().getY()) {
            return circle.radius() - distance;
        } else {
            return circle.radius() + distance;
        }
    }

    private static double getGroundDepth(Circle circle, double y) {
        var depth = y - circle.center().getY() + circle.radius();
        if (depth < 0) {
            return 0.0;
        }
        return depth;
    }

    private static double getGroundAngle(Position groundPosition, int groundIndex, RoomModel roomModel) {
        var angles = new ArrayList<Double>();
        var groundPositionBefore = BattleUtils.getGroundPosition(groundIndex - 1, roomModel);
        if  (groundPositionBefore != null) {
            angles.add(groundPositionBefore.angleTo(groundPosition));
        }
        var groundPositionAfter = BattleUtils.getGroundPosition(groundIndex + 1, roomModel);
        if (groundPositionAfter != null) {
            angles.add(groundPosition.angleTo(groundPositionAfter));
        }
        return angles.stream().reduce(0.0, Double::sum) / angles.size();
    }
}

File: frontend/playground/utils/ground-contact-utils.ts
import {Contact, type Position} from "~/playground/data/common";
import type {RoomModel} from "~/playground/data/model";
import {BattleUtils} from "~/playground/utils/battle-utils";
import {VectorUtils} from "~/playground/utils/vector-utils";
import {type BodyPart, type Circle, HalfCircle, Segment, Trapeze} from "~/playground/data/geometry";
import {GeometryUtils} from "~/playground/utils/geometry-utils";
import {Constants} from "~/playground/data/constants";

export const GroundContactUtils = {

  getContacts(bodyPart: BodyPart, roomModel: RoomModel, withMaxDepth: boolean): Set<Contact>  {
    if (bodyPart instanceof HalfCircle) {
      return GroundContactUtils.getHalfCircleGroundContacts(bodyPart, roomModel, withMaxDepth)
    }
    if (bodyPart instanceof Trapeze) {
      return  GroundContactUtils.getTrapezeGroundContacts(bodyPart, roomModel, withMaxDepth)
    }
    return new Set()
  },

  getHalfCircleGroundContacts(
      halfCircle: HalfCircle,
      roomModel: RoomModel,
      withMaxDepth: boolean
  ): Set<Contact> {
    const groundIndexes = BattleUtils.getGroundIndexesBetween(
        halfCircle.center.x - halfCircle.radius,
        halfCircle.center.x + halfCircle.radius,
        roomModel
    )
    const contacts = new Set<Contact>()

    if (groundIndexes.length === 0) return contacts

    const bottom = halfCircle.chord()
    const halfCircleNormal = VectorUtils.vectorFromTo(
        halfCircle.center,
        BattleUtils.shiftedPosition(halfCircle.center, 1, halfCircle.angle + Math.PI / 2)
    )
    const maxDepth = withMaxDepth ? roomModel.specs.groundMaxDepth : 0

    for (const groundIndex of groundIndexes) {
      const position = BattleUtils.getGroundPosition(groundIndex, roomModel)

      const bottomContact: Contact | null = this.getSegmentGroundContact(bottom,
          position, groundIndex, roomModel, maxDepth, 'HalfCircle bottom with ground')

      let topContact: Contact | null = null
      const distance = BattleUtils.distance(halfCircle.center, position)
      const radiusVector = VectorUtils.vectorFromTo(halfCircle.center, position)

      if (distance < halfCircle.radius && VectorUtils.dotProduct(halfCircleNormal, radiusVector) > 0) {
        const depth = halfCircle.radius - distance - maxDepth
        const contactPosition = BattleUtils.shiftedPosition(halfCircle.center, halfCircle.radius,
            VectorUtils.angleFromTo(halfCircle.center, position))
        topContact = Contact.withAngleUncheckedDepth(
            depth,
            this.getGroundAngle(position, groundIndex, roomModel),
            contactPosition,
            "HalfCircle top with ground"
        )
      }

      let resultContact: Contact | null = null
      if (topContact && bottomContact) {
        resultContact = topContact.depth < bottomContact.depth ? topContact : bottomContact
      } else if (topContact) {
        resultContact = topContact
      } else if (bottomContact) {
        resultContact = bottomContact
      }

      if (resultContact && resultContact.depth >= Constants.INTERPENETRATION_THRESHOLD) {
        contacts.add(resultContact)
      }
    }

    return contacts
  },

  getTrapezeGroundContacts(
      trapeze: Trapeze,
      roomModel: RoomModel,
      withMaxDepth: boolean
  ): Set<Contact> {
    const maxDistance = trapeze.maxDistanceFromCenter()
    const groundIndexes = BattleUtils.getGroundIndexesBetween(
        trapeze.position.x - maxDistance,
        trapeze.position.x + maxDistance,
        roomModel
    )
    const contacts = new Set<Contact>()

    if (groundIndexes.length === 0) {
      return contacts
    }

    const bottom = trapeze.bottom()
    const top = trapeze.top()
    const right = trapeze.right()
    const left = trapeze.left()
    const maxDepth = withMaxDepth ? roomModel.specs.groundMaxDepth : 0

    for (const index of groundIndexes) {
      const position = BattleUtils.getGroundPosition(index, roomModel)

      const bottomContact = this.getSegmentGroundContact(bottom, position, index, roomModel, maxDepth, 'Trapeze bottom with ground')
      const topContact = this.getSegmentGroundContact(top, position, index, roomModel, maxDepth, 'Trapeze top with ground')
      const rightContact = this.getSegmentGroundContact(right, position, index, roomModel, maxDepth, 'Trapeze right with ground')
      const leftContact = this.getSegmentGroundContact(left, position, index, roomModel, maxDepth, 'Trapeze left with ground')

      let resultContact: Contact | null = null

      if (topContact !== null) {
        resultContact = topContact
      }
      if (bottomContact !== null && (resultContact === null || bottomContact.depth < resultContact.depth)) {
        resultContact = bottomContact
      }
      if (rightContact !== null && (resultContact === null || rightContact.depth < resultContact.depth)) {
        resultContact = rightContact
      }
      if (leftContact !== null && (resultContact === null || leftContact.depth < resultContact.depth)) {
        resultContact = leftContact
      }

      if (resultContact !== null && resultContact.depth >= Constants.INTERPENETRATION_THRESHOLD) {
        contacts.add(resultContact)
      }
    }

    return contacts
  },

  getCircleGroundContact(
      circle: Circle,
      roomModel: RoomModel,
      withMaxDepth: boolean
  ): Contact | null {
    const groundSegments = BattleUtils.getGroundSegmentsBetween(
        circle.center.x - circle.radius * 1.1,
        circle.center.x + circle.radius * 1.1,
        roomModel
    )

    if (groundSegments.length === 0) return null

    let nearestPosition: Position | null = null
    let minimalDistance: number | null = null
    let nearestSegment: Segment | null = null

    for (const segment of groundSegments) {
      const position = GeometryUtils.getPointToSegmentProjection(circle.center, segment)
      if (position) {
        const distance = BattleUtils.distance(position, circle.center)

        if (distance < circle.radius &&
            (minimalDistance === null || distance < minimalDistance)) {
          nearestPosition = position
          minimalDistance = distance
          nearestSegment = segment
        }
      }
    }

    if (nearestPosition && minimalDistance !== null && nearestSegment !== null) {
      const contactPosition = BattleUtils.shiftedPosition(circle.center, circle.radius,
          VectorUtils.angleFromTo(circle.center, nearestPosition))
      let depth = BattleUtils.distance(nearestPosition, contactPosition)
      if (withMaxDepth) depth -= roomModel.specs.groundMaxDepth
      const groundAngle = VectorUtils.angleFromTo(nearestSegment.begin, nearestSegment.end)
      return Contact.withAngle(
          depth,
          groundAngle,
          contactPosition
      )
    }

    const fallbackPosition = BattleUtils.getNearestGroundPosition(circle.center.x, roomModel)
    let depth = this.getGroundDepth.fullUnderGround(circle, fallbackPosition.y)
    if (withMaxDepth) depth -= roomModel.specs.groundMaxDepth
    const contactPosition = BattleUtils.shiftedPosition(circle.center, circle.radius, -Math.PI / 2)
    return Contact.withAngle(
        depth,
        0,
        contactPosition
    )
  },

  getSegmentGroundContact(
      segment: Segment,
      groundPosition: Position,
      groundIndex: number,
      roomModel: RoomModel,
      maxDepth: number,
      description: string
  ): Contact | null {
    const segmentPoint = segment.findPointWithX(groundPosition.x)
    if (segmentPoint !== null && segmentPoint.y < groundPosition.y) {
      const projection = GeometryUtils.getPointToSegmentProjection(groundPosition, segment)
      if (projection !== null) {
        const depth = BattleUtils.distance(groundPosition, segmentPoint) - maxDepth
        const normal1 = VectorUtils.vectorFromTo(groundPosition, projection)
        VectorUtils.normalize(normal1)
        const normal2 = VectorUtils.normal(this.getGroundAngle(groundPosition, groundIndex, roomModel))
        VectorUtils.normalize(normal2)
        const normal = VectorUtils.sumOf(normal1, normal2)
        VectorUtils.normalize(normal)
        return Contact.withNormalUncheckedDepth(depth, normal, projection, description)
      }
    }
    return null
  },

  getGroundDepth: {
    fullUnderGround(circle: Circle, y: number): number {
      const depth = y - circle.center.y + circle.radius
      return depth < 0 ? 0 : depth
    }
  },

  getGroundAngle(
      groundPosition: Position,
      groundIndex: number,
      roomModel: RoomModel
  ): number {
    const angles: number[] = []
    const prevPosition = BattleUtils.getGroundPosition(groundIndex - 1, roomModel)
    if (prevPosition) {
      angles.push(VectorUtils.angleFromTo(prevPosition, groundPosition))
    }
    const nextPosition = BattleUtils.getGroundPosition(groundIndex + 1, roomModel)
    if (nextPosition) {
      angles.push(VectorUtils.angleFromTo(groundPosition, nextPosition))
    }
    return angles.length > 0
        ? angles.reduce((sum, angle) => sum + angle, 0) / angles.length
        : 0
  },
}
